unittests/savanna_cluster.hpp

```diff
@@ -0,0 +1,244 @@
+   template<class tester>
+   using finalizer_keys   = eosio::testing::finalizer_keys<tester>;
+
+   class cluster_t;
+
+   // ----------------------------------------------------------------------------
+   class node_t : public tester {
+      uint32_t                prev_lib_num{0};
+      size_t                  node_idx;
+      cluster_t&              cluster;
+      finalizer_keys<tester>  finkeys;
+      size_t                  cur_key{0}; // index of key used in current policy
+
+   public:
+      node_t(size_t node_idx, cluster_t& cluster, setup_policy policy = setup_policy::none);
+
+      void set_node_finalizers(size_t keys_per_node, size_t num_nodes) {
+         finkeys.init_keys(keys_per_node * num_nodes, num_nodes);
+
+         size_t first_node_key = node_idx * keys_per_node;
+         cur_key               = first_node_key;
+         finkeys.set_node_finalizers(first_node_key, keys_per_node);
+      }
+
+      std::pair<std::vector<bls_public_key>, eosio::chain::finalizer_policy>
+      transition_to_savanna(std::span<const size_t> indices) {
+         auto pubkeys = finkeys.set_finalizer_policy(indices).pubkeys;
+         auto policy  = finkeys.transition_to_savanna();
+         return { pubkeys, policy };
+      }
+
+      // returns true if LIB advances on this node since we last checked
+      bool lib_advancing() {
+         if (lib_num() > prev_lib_num) {
+            prev_lib_num = lib_num();
+            return true;
+         }
+         return false;
+      }
+
+      void reset_lib() { prev_lib_num = lib_num(); }
+
+      uint32_t lib_num() const { return lib_block->block_num(); }
+
+      uint32_t head_num() const { return control->fork_db_head_block_num(); }
+
+      void push_blocks(node_t& to, uint32_t block_num_limit = std::numeric_limits<uint32_t>::max()) const {
+         while (to.head_num() < std::min(head_num(), block_num_limit)) {
+            auto sb = control->fetch_block_by_number(to.control->fork_db_head_block_num() + 1);
+            to.push_block(sb);
+         }
+      }
+   };
+
+   // ---------------------------------------------------------------------------------------
+   // cluster_t
+   // ---------
+   //
+   // Set up a test network which consists of 4 nodes, all of which have transitioned to
+   // the Savanna consensus.
+   //
+   // They are all finalizers (Each node has one finalizer) and can all produce blocks.
+   // quorum is computed using the same formula as in the system contracts
+   // (so quorum == 3)
+   //
+   //
+   // By default they are all connected, receive all produced blocks, vote on them,
+   // and send their votes to all other nodes.
+   //
+   // It is possible to split the 'virtual' network using `cluster_t::split()`.
+   //  --------------------------------------------------------------------------------------
+   class cluster_t {
+   public:
+      static constexpr size_t num_nodes = 4;
+      static constexpr size_t keys_per_node = 10;
```

Comment:
Well, I did make the `finality_test_cluster` a template class which takes `NUM_NODES` as a template parameter. I felt that it was less needed for the `savanna_cluster`, but yes probably  I didn't envision all the possible use cases so I'll change it. 

Do you think there should be a minimal number of nodes? With one node you don't create branches and finality always advances. With two or 3 nodes you need all the nodes to vote for finality to advance. 4 seemed like a nice number where the quorum is 3. 

As of now, `finality_test_cluster` requires a minimum of 4 nodes.

Reply:
> It would be great if the number of nodes were dynamic and determined by an argument to the cluster. 

Would you mind if I made the cluster a `template<size_t NUM_NODES> ` instead of the cluster taking an argument? It would allow to keep the array of nodes as a `std::array`.

Reply:
OK, made `savanna_cluster` a template:

```
template <size_t NUM_NODES, size_t KEYS_PER_NODE = 10> requires (NUM_NODES > 3)
class cluster_t {
...
```

Reply:
I don't see a strong reason to advocate for using a `std::array` over a `std::vector` in this case. It is just one allocation during the constructor, and performance of an extra indirection doesn't really matter in this case.

And on the other hand, I think it makes the code a little more complicated by making it a template. In particular, making the the node class into a template so that it can take the now template cluster class.

By the way the constructor of the cluster is still assuming exactly four nodes despite making `NUM_NODES` a template paramater. I think it is fine to enforce a minimum value of 4 for `num_nodes`. But I do think this would all be a little cleaner if it was just provided as a constructor argument in a class that is not a template.

Also, I do not see why `keys_per_node` is required at all? In fact, I find the `finalizer_keys` template class to be hard to follow in general. So far, I am not seeing why you need more than one finalizer key per node. I can see how there could be value for that for some tests where you want to rotate the finalizer key for a node. But even then I don't see why you would need to initial a fixed number of finalizer per node rather than just generation finalizer keys as needed.

For example, https://github.com/AntelopeIO/spring/blob/main/libraries/testing/include/eosio/testing/bls_utils.hpp could be modified to take a `sha256` hash directly as the seed to generate the private key, public key, and proof of possession rather than arbitrarily limiting oneself only to a `eosio::chain::name`. Then the tester utility could generate a unique key (within the cluster) by tracking a finalizer key sequence number per each node and generating a `sha256` hash to use as the seed from a SHA256 hash of both the node index and the finalizer key sequence number.

Reply:
A problem is that when using `BOOST_FIXTURE_TEST_CASE `, you provide a parent class, but as far as I can tell you cannot pass an argument to the class constructor. So I feel that if we want to support a variable number of nodes, the template is actually a better solution here.

Reply:
> By the way the constructor of the cluster is still assuming exactly four nodes despite making `NUM_NODES` a template paramater.

You are right, I may have to switch to a vector anyways to initialize the nodes correctly.
I'll do that in the other branch `gh_10`, because if I switch to a vector, I cannot have the references `node0`, `node1` declared as member in the class, and I'll remove these in `gh_10`.

Reply:
> A problem is that when using BOOST_FIXTURE_TEST_CASE , you provide a parent class, but as far as I can tell you cannot pass an argument to the class constructor. So I feel that if we want to support a variable number of nodes, the template is actually a better solution here.

A typical way of handling this is to create a wrapper derived class which uses its default constructor to call the base constructor with different values.

For convenience, the `savanna_cluster` could have default values for the parameters of its constructor so that it can be used directly in most tests that do not need to change the number of nodes.

> You are right, I may have to switch to a vector anyways to initialize the nodes correctly.
I'll do that in the other branch gh_10, because if I switch to a vector, I cannot have the references node0, node1 declared as member in the class, and I'll remove these in gh_10.

Yes, I think the `node0`, `node1`, etc. should be removed. I would rather see the incorrect initialization fixed in this PR since as the code currently exists it does not actually work correctly with `NUM_NODES` other than 4.

Reply:
OK, doing it. There is the same pattern (`template<size_t NUM_NODES>`, `node0` and `node1` aliases) in `finality_test_cluster.hpp`. Should I update this one as well for consistency?

Reply:
I updated `savanna_cluster` to not be a template anymore, use a vector for nodes and forego the aliases.